sdk/include/opentelemetry/sdk/http/http_client.h

```diff
@@ -0,0 +1,139 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include "opentelemetry/http/http_client.h"
+#include "opentelemetry/version.h"
+
+#include <map>
+#include <vector>
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace http
+{
+namespace http_api = opentelemetry::http;
+
+// HttpHeaders implementation
+class HttpHeaders : http_api::HttpHeaders, std::multimap<std::string, std::string>
+{
+
+public:
+  virtual void set(nostd::string_view const &name, nostd::string_view const &value) override
+  {
+    auto range = equal_range(std::string(name.data()));
+    auto hint  = erase(range.first, range.second);
+    insert(hint, std::make_pair(name, value));
+  }
+
+  virtual void add(nostd::string_view const &name, nostd::string_view const &value) override
+  {
+    insert(std::make_pair(name, value));
+  }
+
+  virtual nostd::string_view const &get(nostd::string_view const &name) const override
+  {
+    auto it = find(std::string(name.data()));
+    return (it != end()) ? it->second : m_empty;
+  }
+
+  virtual bool has(nostd::string_view const &name) const override
+  {
+    auto it = find(std::string(name.data()));
+    return (it != end());
+  }
+
+private:
+  nostd::string_view m_empty{};
+};
+
+class SimpleHttpRequest : public http_api::HttpRequest
+{
+
+public:
+  SimpleHttpRequest(nostd::string_view const &id) : id_(id), method_("GET") {}
+
+  // Gets the HTTP request ID.
+  virtual const nostd::string_view &GetId() const override { return id_; }
+
+  // Sets the request method
+  virtual void SetMethod(nostd::string_view const &method) override { method_ = method; }
+
+  // Sets the HTTP request URI.
+  virtual void SetUrl(nostd::string_view const &url) override { url_ = url; }
+
+  // Gets the HTTP request headers.
+  virtual http_api::HttpHeaders *GetHeaders() const override { return headers_; }
+
+  // Sets the request body.
+  virtual void SetBody(const uint8_t *const body, const size_t len) override
+  {
+    for (int i = 0; i < len; i++)
+    {
+      body_.push_back(body[i]);
+    }
+  }
+
+  virtual void GetBody(uint8_t *body, size_t &len) override
```

Comment:
I'd suggest we refactor this to use `std::vector<uint8_t>` instead of raw pointer. This method is also unsafe, as presented, because we never know if buffer at `*body` is large enough to fit the contents of `body_`.

Reply:
The reason to use raw pointer was to avoid stl libraries in API as per the guidelines. Although agree that this method is unsafe in current form.

Reply:
We only have to avoid STL in the API part, not in the SDK.

However, the best solution here would be to pass a `nostd::span<uint8_t>`.

Reply:
@pyohannes  @maxgolov  - `http_sdk::SimpleHttpRequest::GetBody` actually overrides `http_api::HttpRequest::GetBody` pure virtual function. So using `std::vector<uint8_t>` or `nostd::span<uint8_t>` in SDK would means I need to change it in API too, which is what I am trying to avoid. Or am I missing something in the comment here ?

Reply:
I'm not sure we need to expose this at API layer. Both - HTTP client interface and implementation may be part of SDK, and/or exporter compiled into SDK, not in API. Not passing anything across ABI boundary. Yes, the change from pointer to container reference has to be made in both classes.. I think it'd be easier to deal with a reference to owning `std::vector` rather than non-owning `nostd::span`. That way the client may be easily implemented in older compiler without taking dependency on `nostd` classes, e.g. Chromium HTTP stack - `C++14`.. Reason why I'm mentioning this, is we may swiftly contribute the existing code `(c) Microsoft - Apache OR MIT` for the following HTTP implementations:
- WinInet
- libcurl
- cpp rest SDK
- Chromium HTTP stack
- Apple HTTP stack
- bindings to Android reverse-JNI HTTP stack
And all of these implementations won't have to take dependency on `nostd` (which I think we should reserve only for ABI-stable interfaces, not unnecessarily using elsewhere).

Reply:
All the raw pointers would be converted to `shared_ptr`/`unique_ptr` in new PR.